File: experiments_realworld/graphs/problem02/implementations/optimal/solution.py
# CSES 1197: Cycle Finding
# Bellman-Ford Negative Cycle Detection - O(nm)

import sys

def main():
    data = sys.stdin.buffer.read().split()
    it = iter(data)
    n = int(next(it))
    m = int(next(it))

    edges = []
    for i in range(m):
        a = int(next(it))
        b = int(next(it))
        w = int(next(it))
        edges.append((a, b, w))

    dist = [0] * (n + 1)          # Implicit super-source
    parent = [-1] * (n + 1)

    x = -1
    
    # Bellman-Ford: n iterations of edge relaxation
    for _ in range(n):
        x = -1
        for a, b, w in edges:
            nb = dist[a] + w
            if nb < dist[b]:
                dist[b] = nb
                parent[b] = a
                x = b

    if x == -1:   
        print("NO")
        return

    # Find node guaranteed to be in negative cycle
    y = x
    for _ in range(n):
        y = parent[y]

    # Reconstruct cycle and print closed form
    cycle = []
    v = y
    while True:
        cycle.append(v)
        v = parent[v]
        if v == y:
            cycle.append(v)  # Close the cycle
            break
    cycle.reverse()

    print("YES")
    print(*cycle)

if __name__ == "__main__":
    main()

File: experiments_realworld/graphs/problem02/implementations/optimal/solution.cpp
// CSES 1197: Cycle Finding
// Bellman-Ford Negative Cycle Detection - O(nm)

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int n, m;
    if (!(cin >> n >> m)) return 0;

    struct Edge {
        int a, b;
        long long c;
    };
    
    vector<Edge> edges(m);
    for (int i = 0; i < m; ++i) {
        cin >> edges[i].a >> edges[i].b >> edges[i].c;
    }

    vector<long long> dist(n + 1, 0);   // Implicit super-source
    vector<int> parent(n + 1, -1);

    int x = -1;
    
    // Bellman-Ford: n iterations of edge relaxation
    for (int i = 0; i < n; ++i) {
        x = -1;
        for (auto &e : edges) {
            if (dist[e.a] + e.c < dist[e.b]) {
                dist[e.b] = dist[e.a] + e.c;
                parent[e.b] = e.a;
                x = e.b;
            }
        }
    }

    if (x == -1) {
        cout << "NO\n";
        return 0;
    }

    // Find node guaranteed to be in negative cycle
    int y = x;
    for (int i = 0; i < n; ++i) {
        y = parent[y];
    }

    // Reconstruct cycle
    vector<int> cycle;
    for (int v = y;; v = parent[v]) {
        cycle.push_back(v);
        if (v == y && cycle.size() > 1) break;
    }
    reverse(cycle.begin(), cycle.end());

    cout << "YES\n";
    for (size_t i = 0; i < cycle.size(); ++i) {
        if (i) cout << ' ';
        cout << cycle[i];
    }
    cout << "\n";
    
    return 0;
}
